vocode/streaming/streaming_conversation.py

```diff
@@ -880,83 +885,104 @@ async def send_speech_to_output(
-        cut_off = False
-        chunk_size = self._get_synthesizer_chunk_size(seconds_per_chunk)
-        chunk_idx = 0
-        seconds_spoken = 0.0
         logger.debug(f"Start sending speech {message} to output")
 
         first_chunk_span = self._maybe_create_first_chunk_span(synthesis_result, message)
-        chunk_queue: asyncio.Queue[Optional[SynthesisResult.ChunkResult]] = asyncio.Queue()
-        get_chunks_task = asyncio_create_task_with_done_error_log(
-            get_chunks(chunk_queue, synthesis_result.chunk_generator),
-        )
-        first = True
-        while True:
-            chunk_result = await chunk_queue.get()
-            if chunk_result is None:
-                break
-            if first and first_chunk_span:
-                self._track_first_chunk(first_chunk_span, synthesis_result)
-            first = False
-            start_time = time.time()
-            speech_length_seconds = seconds_per_chunk * (len(chunk_result.chunk) / chunk_size)
-            seconds_spoken = chunk_idx * seconds_per_chunk
+        audio_chunks: List[AudioChunk] = []
+        processed_events: List[asyncio.Event] = []
+        interrupted_before_all_chunks_sent = False
+        async for chunk_idx, chunk_result in enumerate_async_iter(synthesis_result.chunk_generator):
             if stop_event.is_set():
-                logger.debug(
-                    "Interrupted, stopping text to speech after {} chunks".format(chunk_idx),
-                )
-                message_sent = synthesis_result.get_message_up_to(seconds_spoken)
-                cut_off = True
+                logger.debug("Interrupted before all chunks were sent")
+                interrupted_before_all_chunks_sent = True
                 break
-            if chunk_idx == 0:
-                if started_event:
-                    started_event.set()
-            self.output_device.consume_nonblocking(chunk_result.chunk)
-            end_time = time.time()
-            await asyncio.sleep(
-                max(
-                    speech_length_seconds
-                    - (end_time - start_time)
-                    - self.per_chunk_allowance_seconds,
-                    0,
-                ),
+            processed_event = asyncio.Event()
+            audio_chunk = AudioChunk(
+                data=chunk_result.chunk,
             )
-            self.mark_last_action_timestamp()
-            chunk_idx += 1
-            seconds_spoken += seconds_per_chunk
-            if transcript_message:
-                transcript_message.text = synthesis_result.get_message_up_to(seconds_spoken)
-        get_chunks_task.cancel()
+            # register callbacks
+            setattr(audio_chunk, "on_play", create_on_play_callback(chunk_idx, processed_event))
+            setattr(
+                audio_chunk,
+                "on_interrupt",
+                create_on_interrupt_callback(processed_event),
+            )
+            async with self.interrupt_lock:
```

Comment:
why is the interrupt lock required here?

Reply:
if we send a chunk while an interruption is inflight, you could get the following two scary scenarios:

(1)
- a chunk is marked interrupted
- _we still send the chunk to the output device_
- then we send the clear message

(2)
- a chunk is marked interrupted
- we send the clear message
- _we still send the chunk to the output device_